<a href="https://colab.research.google.com/github/Meera-Ajay-25/Machine-Learning-projects/blob/main/porter_regression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Project 2: **Porter delivery time estimation**

Algorithm Used: **Multiple regression**

Steps:


*   Loading the dataset using pandas library
*   preprocessing the dataset using Excel.


*   Splitting the dataset
*   Model building using Multiple regression Algorithm


*   Model testing using MSE,MAE,R2-SCORE,RMSE










In [19]:
#Loading the dataset using python library
import pandas as pd
df=pd.read_csv("/content/porter1.csv")
df

market_id  created_at  actual_delivery_time  \
0             2.0    60521752              60521754   
1             2.0    60523079              60523082   
2             2.0    60543321              60543326   
3             2.0    60530441              60530447   
4             2.0    60522084              60522090   
...           ...         ...                   ...   
197413        2.0    60542167              60542823   
197414        2.0    60552257              60553019   
197415        3.0    60537731              60538534   
197416        1.0    60553591              60554498   
197417        2.0    60531978              60532928   

                                store_id store_primary_category  \
0       c56a022b15250525f8b9bdfc41a13152                mexican   
1       aa2a77371374094fe9e0bc1de3f94ed9               sandwich   
2       d43ab110ab2489d6b9b2caa394bf920f                mexican   
3       1afa34a7f984eeabdbb0a7d494132ee5                 indian   
4       509421719867b607d48a0a1b0ae35c89               sandwich   
...                                  ...                    ...   
197413  5fc7c9bd1fcb12799f02da8adfa4954f                alcohol   
197414  9380e398ee9bea45b992a3daaa6b7c4d                  pizza   
197415  831b342d8a83408e5960e9b0c5f31f0c                   thai   
197416  b132ecc1609bfcf302615847c1caa69a                 indian   
197417  c1502ae5a4d514baec129f72948c266e                 burger   

        order_protocol  total_items  subtotal  num_distinct_items  \
0                  2.0            3      1455                   2   
1                  2.0            1      1195                   1   
2                  5.0            1      1499                   1   
3                  3.0            6      1950                   5   
4                  2.0            7      2527                   4   
...                ...          ...       ...                 ...   
197413             5.0            3      2850                   3   
197414             6.0            1       990                   1   
197415             2.0            3      2185                   3   
197416             3.0            4      3660                   4   
197417             4.0            3      2379                   2   

        min_item_price  Average_rate  max_item_price  total_onshift_partners  \
0                  355         550.0             745                    44.0   
1                 1195        1195.0            1195                    60.0   
2                  799         799.0             799                    67.0   
3                  100         575.0            1050                    52.0   
4                   69         434.0             799                    51.0   
...                ...           ...             ...                     ...   
197413             200         850.0            1500                    96.0   
197414             795         795.0             795                   114.0   
197415             495         745.0             995                    19.0   
197416             375         785.0            1195                    71.0   
197417             389         542.0             695                   109.0   

        total_busy_partners  total_outstanding_orders  Actual_time  
0                      48.0                      40.0            2  
1                      74.0                      51.0            3  
2                      58.0                      61.0            5  
3                      54.0                      63.0            6  
4                      63.0                      52.0            6  
...                     ...                       ...          ...  
197413                103.0                     156.0          656  
197414                112.0                     184.0          762  
197415                 19.0                      16.0          803  
197416                 70.0                     111.0         

In [21]:

df.drop(['market_id','created_at','actual_delivery_time','store_id','store_primary_category','min_item_price','max_item_price'],axis=1,inplace=True)

In [22]:
df.dtypes

order_protocol              float64
total_items                   int64
subtotal                      int64
num_distinct_items            int64
Average_rate                float64
total_onshift_partners      float64
total_busy_partners         float64
total_outstanding_orders    float64
Actual_time                   int64
dtype: object

In [26]:
df.isna().sum()

order_protocol              0
total_items                 0
subtotal                    0
num_distinct_items          0
Average_rate                0
total_onshift_partners      0
total_busy_partners         0
total_outstanding_orders    0
Actual_time                 0
dtype: int64

In [24]:
lst=['order_protocol','total_onshift_partners','total_busy_partners','total_outstanding_orders']
for i in lst:
  df[i]=df[i].fillna(df[i].mode()[0])

Preprocessing has been done using Excel sheet. The actual dataset contained two attributes **created_at** and **actual_delivery_time** which was in yyyy-mm-dd,HH:MM:SS format which was converted to minutes using Excel and the difference of both were taken to find the target column **Actual_time**. The Average of **min_item_price** and **max_item_price** were taken and the attribute **Average_rate** was found. Attributes **market_id**,**created_at**,**actual_delivery_time**,**store_id**,**store_primary_category**,**min_item_price**,**max_item_price** were dropped. Missing values in the columns were replaced using mode.

In [25]:
df

order_protocol  total_items  subtotal  num_distinct_items  \
0                  2.0            3      1455                   2   
1                  2.0            1      1195                   1   
2                  5.0            1      1499                   1   
3                  3.0            6      1950                   5   
4                  2.0            7      2527                   4   
...                ...          ...       ...                 ...   
197413             5.0            3      2850                   3   
197414             6.0            1       990                   1   
197415             2.0            3      2185                   3   
197416             3.0            4      3660                   4   
197417             4.0            3      2379                   2   

        Average_rate  total_onshift_partners  total_busy_partners  \
0              550.0                    44.0                 48.0   
1             1195.0                    60.0                 74.0   
2              799.0                    67.0                 58.0   
3              575.0                    52.0                 54.0   
4              434.0                    51.0                 63.0   
...              ...                     ...                  ...   
197413         850.0                    96.0                103.0   
197414         795.0                   114.0                112.0   
197415         745.0                    19.0                 19.0   
197416         785.0                    71.0                 70.0   
197417         542.0                   109.0                102.0   

        total_outstanding_orders  Actual_time  
0                           40.0            2  
1                           51.0            3  
2                           61.0            5  
3                           63.0            6  
4                           52.0            6  
...                          ...          ...  
197413                     156.0          656  
197414                     184.0          762  
197415                      16.0          803  
197416                     111.0          907  
197417                     163.0          950  

[197418 rows x 9 columns]

In [27]:
X=df.iloc[:,:-1]
y=df.iloc[:,-1]

In [28]:
#scaling the dataset using MinMaxScaler
from sklearn.preprocessing import MinMaxScaler
ms=MinMaxScaler()
X_sc=ms.fit_transform(X)

In [29]:
#Performing train_test_split
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X_sc,y,test_size=0.3,random_state=0)

In [42]:
X_train.shape

(138192, 8)

In [43]:
y_train.shape

(138192,)

In [30]:
#model building using Multiple regression
from sklearn.linear_model import LinearRegression
mlr=LinearRegression()
mlr.fit(X_train,y_train)

LinearRegression()

In [31]:
y_pred=mlr.predict(X_test)
y_pred

array([45.97546569, 48.36210813, 64.18481564, ..., 47.37689494,
       41.85417192, 49.32127187])

In [32]:
y_test.values

array([34, 62, 47, ..., 48, 45, 49])

In [37]:
#Comparing Actual and predicted Values
Result=pd.DataFrame({'Actual Values':y_test,'Predicted Values':y_pred})
Result

Actual Values  Predicted Values
42623              34         45.975466
163836             62         48.362108
114029             47         64.184816
175729             69         42.711898
110050             47         47.022399
...               ...               ...
2603               20         43.307347
187067             80         62.857953
116988             48         47.376895
102029             45         41.854172
120862             49         49.321272

[59226 rows x 2 columns]

In [36]:
print('coefficient:',mlr.coef_)
print('slope:',mlr.intercept_)

coefficient: [ -4.33570386 -19.66829492  43.42443617  10.63415051  19.22964502
 -63.17169641 -21.09130172 103.71725077]
slope: 43.574688964800316


In [38]:
#model validation
from sklearn.metrics import mean_absolute_error,mean_squared_error, r2_score
import numpy as np
print("mean absolute error:",mean_absolute_error(y_test,y_pred))
print("mean squared error:",mean_squared_error(y_test,y_pred))
print("root mean squared error:",np.sqrt(mean_squared_error(y_test,y_pred)))
print("r2-score:",r2_score(y_test,y_pred))

mean absolute error: 12.465569566950341
mean squared error: 315.8295149976578
root mean squared error: 17.771592922348233
r2-score: 0.14186419223236657


In [45]:
import warnings
warnings.filterwarnings('ignore')
y_new=mlr.predict(ms.transform([[5,1,3660,1,600,52,48,40]]))
y_new

array([36.48359567])

After Model building the results obtained are

mean absolute error: 12.465569566950341
mean squared error: 315.8295149976578
root mean squared error: 17.771592922348233
r2-score: 0.14186419223236657

predicting with separate values [5,1,3660,1,600,52,48,40] we get the delivery time as Approximately **36 minutes** 